### PR TITLE
OS-8280 Add /proc/sys/kernel/random/uuid to LX brands

### PR DESCRIPTION
**Pull request by pekdon**
/proc/sys/kernel/random/uuid generates a random UUID on each read and is used by software such as xapian.

### CODE REVIEW

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (citrus-it) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5003]: This line is no-longer necessary, it can be removed if you wish.

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (citrus-it) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5020]: How about adding comments here to make it obvious what these lines so?

```suggestion
	/* Set UUID version to 4 (random) */
	r[6] = 0x40 | (r[6] & 0x0f);
	/* Set UUID variant to 1 */
	r[8] = 0x80 | (r[8] & 0x3f);
```

**Review by citrus-it — COMMENTED** [review]
As an interested third party, I have a couple of comments on this, and thanks for doing it!

Since there are now two places in this file which generate a random UUID, and the other place doesn't generate a valid UUID, how about making a new function that generates a valid type 4 UUID, then use that function for both this and `random/boot_id` ? Real linux does generate valid type 4 UUIDs for `boot_id` so it would be nice if we did the same.

**Comment by pekdon** [comment]
Don't know why I didn't see these comments earlier, anyway, will address them tomorrow.

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (citrus-it) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:4935]: `random_get_bytes()` can possibly fail if there is insufficient entropy, returning EAGAIN.
In that case, the UUID is going to be built from whatever was on the stack.
I think we should rather just fall back to `random_get_pseudo_bytes()` so it's at least a bit more random.

```suggestion
	if (random_get_bytes(r, sizeof (r)) != 0)
	        (void) random_get_pseudo_bytes(r, sizeof (r));
```

Comment (pekdon) [reply usr/src/uts/common/brand/lx/procfs/lx_prvnops.c]: Thanks for the feedback, falsely assumed it wouldn't fail as the original code didn't check the return value either.

**Comment by danmcd** [comment]
Can you leave any notes about how you tested this?  I'd like to pull this in, but I can't w/o knowing how you tested this (so we can repeat it internally).  The code looks good and I'll mark it as a +1 for review, but we can't grant integration approval without testing notes.

**Review by danmcd — APPROVED** [review]
Needs test notes, but code seems decent enough.

**Comment by pekdon** [comment]
> Can you leave any notes about how you tested this? I'd like to pull this in, but I can't w/o knowing how you tested this (so we can repeat it internally). The code looks good and I'll mark it as a +1 for review, but we can't grant integration approval without testing notes.

So far I've only done manual tests being all new to SmartOS and it's ecosystem. What I've done is to ensure that the output is valid version 4 variant 1 UUIDs using online validators, and verified that boot_id does not change. The code is yet to be stress tested.

While reading up on testing in the SmartOS project I noticed the lx-tests repository so will create a PR which add simple functional tests there.

Any specific tests other than writing functional tests in lx-tests you would want me to do @danmcd?

**Comment by danmcd** [comment]
Honestly, for this, you probably don't need to worry about stress-tests.  I was more concerned with was it any different from expectations (i.e. would a Linux user not be surprised by a difference) and it seems it hasn't.  You also regression-tested boot_id, which is also good.

@citrus-it told me via chat he's done your tests on his own integration of this into OmniOS, so I'd like him to report here, and I'll turn this into an OS ticket as well.

**Comment by danmcd** [comment]
OS ticket filed:  https://smartos.org/bugview/OS-8280
Once Andy's test notes are in (and I may need to cstyle this according to him?) either I or someone else can IA this.

**Comment by pekdon** [comment]
Have some simple tests prepared now, https://github.com/pekdon/lx-tests/tree/procfs-random-tests, but I guess those should wait until this is merged.

**Comment by pekdon** [comment]
> Honestly, for this, you probably don't need to worry about stress-tests. 

The only thing that would be nice to know that is that it gracefully falls back into using pseudorandom code on low entropy situations I guess, hard to device such a test however.

**Comment by citrus-it** [comment]
I integrated this into OmniOS and did some basic testing - booting an LX zone and reading from the new `uuid` file as fast as possible, then checked the returned UUIDs were valid type 4. Even when reading as fast as possible, I wasn't able to observe `random_get_bytes()` failing so that it needed to fall back to `random_get_pseudo_bytes()` though.

I also check that the `boot_id` file returned a valid type 4 UUID and it did not change over time, and that it changed on zone reboot.

Re the cstyle, it is just the indentation of the continuation lines to `snprintf()` which should be 4 spaces if SmartOS is aligned with the illumos style.

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (danmcd) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:4944]: Yeah, please fix this to be tab-plus-4-spaces?  Once that's done I'm ready to IA or encourage @papertigers to IA.

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (papertigers) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5007]: In smartos lx proc can be found at `/system/lxproc` in a native zone.  I don't see the sys dir under that path, but I do see it in an lx zone.  I just want to confirm someone in a native zone cant use this path to trigger this assert. I would like to verify that we don't mount sys in native.

Comment (papertigers) [reply usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5007]: https://github.com/joyent/illumos-joyent/blob/2a7c46f8d5561b2947bce41d61abe7aae999ba8e/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c#L31

I guess this answers my question :)

**Review by papertigers — DISMISSED** [review]

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (papertigers) [inline usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5004]: I have given this a +1 
But one nit if you feel like addressing is we could probably pull in `sys/uuid.h` and use `UUID_PRINTABLE_STRING_LENGTH`  from there instead of `LX_BOOTID_LEN`

Comment (pekdon) [reply usr/src/uts/common/brand/lx/procfs/lx_prvnops.c:5004]: Sure, can fix that. Busy now but will try to find some time before the end of the day here.

**Comment by danmcd** [comment]
Unless @pekdon wants to fix that last UUID-length nit, I'm willing to IA this given the testing it's got.

File: usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
Comment (pekdon) [reply usr/src/uts/common/brand/lx/procfs/lx_prvnops.c]: Removed LX_BOOTID_LEN and using UUID_PRINTABLE_STRING_LENGTH instead now.

**Review by papertigers — APPROVED** [review]

**Pull request merged**
